timeout doesn't work with AppEngineAdapter

hey again! so, it looks like timeout may not be making it through AppEngineAdapter. requests always times out in 5s, regardless of what i pass as timeout, in both dev_appserver and prod.

repro code and example traceback below. other kwargs (eg headers) work ok. cc @mikelambert. thanks in advance!

```py
import requests
from requests_toolbelt.adapters import appengine
appengine.monkeypatch()
requests.get('http://httpbin.org/delay/10', timeout=999, headers={'User-Agent': 'xyz'})
```

```
 Traceback (most recent call last):
File "/base/data/home/apps/s~brid-gy/7.392575605249396594/tasks.py", line 521, in do_send_webmentions
  if not mention.send(timeout=999, headers=util.USER_AGENT_HEADER):
File "/base/data/home/apps/s~brid-gy/7.392575605249396594/local/lib/python2.7/site-packages/webmentiontools/send.py", line 27, in send
  return self._notifyReceiver()
File "/base/data/home/apps/s~brid-gy/7.392575605249396594/local/lib/python2.7/site-packages/webmentiontools/send.py", line 79, in _notifyReceiver
  r = requests.post(self.receiver_endpoint, verify=False, data=payload, **self.requests_kwargs)
File "/base/data/home/apps/s~brid-gy/7.392575605249396594/local/lib/python2.7/site-packages/requests/api.py", line 111, in post
  return request('post', url, data=data, json=json, **kwargs)
File "/base/data/home/apps/s~brid-gy/7.392575605249396594/local/lib/python2.7/site-packages/requests/api.py", line 57, in request
  return session.request(method=method, url=url, **kwargs)
File "/base/data/home/apps/s~brid-gy/7.392575605249396594/local/lib/python2.7/site-packages/requests/sessions.py", line 475, in request
  resp = self.send(prep, **send_kwargs)
File "/base/data/home/apps/s~brid-gy/7.392575605249396594/local/lib/python2.7/site-packages/requests/sessions.py", line 585, in send
  r = adapter.send(request, **kwargs)
File "/base/data/home/apps/s~brid-gy/7.392575605249396594/local/lib/python2.7/site-packages/requests/adapters.py", line 403, in send
  timeout=timeout
File "/base/data/home/apps/s~brid-gy/7.392575605249396594/local/lib/python2.7/site-packages/requests_toolbelt/adapters/appengine.py", line 117, in urlopen
  **response_kw)
File "/base/data/home/apps/s~brid-gy/7.392575605249396594/local/lib/python2.7/site-packages/requests/packages/urllib3/contrib/appengine.py", line 109, in urlopen
  raise TimeoutError(self, e)
TimeoutError: (<requests.packages.urllib3.contrib.appengine.AppEngineManager object at 0xfb110df0>, DeadlineExceededError('Deadline exceeded while waiting for HTTP response from URL: http://www.daily-pia.de/?webmention=endpoint',))
```

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/34001023-timeout-doesn-t-work-with-appengineadapter?utm_campaign=plugin&utm_content=tracker%2F418367&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F418367&utm_medium=issues&utm_source=github).
</bountysource-plugin>

btw, i started to tackle this myself, but got a wall of intimidating failures when i ran the tests, enough different kinds that i just backed away slowly. let me know if it's something obvious, happy to try again!

```
GLOB sdist-make: /Users/ryan/src/requests-toolbelt/setup.py
py27 inst-nodeps: /Users/ryan/src/requests-toolbelt/.tox/dist/requests-toolbelt-0.6.1.zip
py27 installed: betamax==0.7.0,funcsigs==1.0.2,mock==2.0.0,pbr==1.9.1,py==1.4.31,pytest==2.9.1,requests==2.10.0,requests-toolbelt==0.6.1,six==1.10.0
py27 runtests: PYTHONHASHSEED='3594687790'
py27 runtests: commands[0] | py.test
sE...F......EEEEE...........E...........ssss.......F.........s..............sEsssF............................................................................................................................................
==================================== ERRORS ====================================
 ERROR collecting local/lib/python2.7/site-packages/pbr/tests/test_commands.py _
.tox/py27/lib/python2.7/site-packages/py/_path/local.py:650: in pyimport
    __import__(modname)
local/lib/python2.7/site-packages/pbr/tests/__init__.py:16: in <module>
    import testscenarios
E   ImportError: No module named testscenarios
__ ERROR collecting local/lib/python2.7/site-packages/pbr/tests/test_core.py ___
.tox/py27/lib/python2.7/site-packages/py/_path/local.py:650: in pyimport
    __import__(modname)
local/lib/python2.7/site-packages/pbr/tests/__init__.py:16: in <module>
    import testscenarios
E   ImportError: No module named testscenarios
__ ERROR collecting local/lib/python2.7/site-packages/pbr/tests/test_files.py __
.tox/py27/lib/python2.7/site-packages/py/_path/local.py:650: in pyimport
    __import__(modname)
local/lib/python2.7/site-packages/pbr/tests/__init__.py:16: in <module>
    import testscenarios
E   ImportError: No module named testscenarios
__ ERROR collecting local/lib/python2.7/site-packages/pbr/tests/test_hooks.py __
.tox/py27/lib/python2.7/site-packages/py/_path/local.py:650: in pyimport
    __import__(modname)
local/lib/python2.7/site-packages/pbr/tests/__init__.py:16: in <module>
    import testscenarios
E   ImportError: No module named testscenarios
 ERROR collecting local/lib/python2.7/site-packages/pbr/tests/test_integration.py 
.tox/py27/lib/python2.7/site-packages/py/_path/local.py:650: in pyimport
    __import__(modname)
local/lib/python2.7/site-packages/pbr/tests/__init__.py:16: in <module>
    import testscenarios
E   ImportError: No module named testscenarios
 ERROR collecting local/lib/python2.7/site-packages/pbr/tests/test_packaging.py 
.tox/py27/lib/python2.7/site-packages/py/_path/local.py:650: in pyimport
    __import__(modname)
local/lib/python2.7/site-packages/pbr/tests/__init__.py:16: in <module>
    import testscenarios
E   ImportError: No module named testscenarios
__ ERROR collecting local/lib/python2.7/site-packages/pbr/tests/test_setup.py __
.tox/py27/lib/python2.7/site-packages/py/_path/local.py:650: in pyimport
    __import__(modname)
local/lib/python2.7/site-packages/pbr/tests/__init__.py:16: in <module>
    import testscenarios
E   ImportError: No module named testscenarios
__ ERROR collecting local/lib/python2.7/site-packages/pbr/tests/test_util.py ___
.tox/py27/lib/python2.7/site-packages/py/_path/local.py:650: in pyimport
    __import__(modname)
local/lib/python2.7/site-packages/pbr/tests/__init__.py:16: in <module>
    import testscenarios
E   ImportError: No module named testscenarios
_ ERROR collecting local/lib/python2.7/site-packages/pbr/tests/test_version.py _
.tox/py27/lib/python2.7/site-packages/py/_path/local.py:650: in pyimport
    __import__(modname)
local/lib/python2.7/site-packages/pbr/tests/__init__.py:16: in <module>
    import testscenarios
E   ImportError: No module named testscenarios
__ ERROR collecting local/lib/python2.7/site-packages/pbr/tests/test_wsgi.py ___
.tox/py27/lib/python2.7/site-packages/py/_path/local.py:650: in pyimport
    __import__(modname)
local/lib/python2.7/site-packages/pbr/tests/__init__.py:16: in <module>
    import testscenarios
E   ImportError: No module named testscenarios
 ERROR collecting local/lib/python2.7/site-packages/pkg_resources/tests/test_pkg_resources.py 
.tox/py27/lib/python2.7/site-packages/py/_path/local.py:650: in pyimport
    __import__(modname)
E   ImportError: No module named tests.test_pkg_resources
 ERROR collecting local/lib/python2.7/site-packages/pkg_resources/tests/test_resources.py 
.tox/py27/lib/python2.7/site-packages/py/_path/local.py:650: in pyimport
    __import__(modname)
E   ImportError: No module named tests.test_resources
____________________ ERROR at setup of Test.test_bdist_egg _____________________
file /Users/ryan/src/requests-toolbelt/local/lib/python2.7/site-packages/setuptools/tests/test_bdist_egg.py, line 29
      def test_bdist_egg(self, setup_context, user_override):
        fixture 'user_override' not found
        available fixtures: record_xml_property, recwarn, betamax_recorder, tmpdir, cache, monkeypatch, pytestconfig, setup_context, tmpdir_factory, betamax_session, capsys, capfd
        use 'py.test --fixtures [testpath]' for help on them.

/Users/ryan/src/requests-toolbelt/local/lib/python2.7/site-packages/setuptools/tests/test_bdist_egg.py:29
_______ ERROR at setup of TestUserInstallTest.test_user_install_implied ________
file /Users/ryan/src/requests-toolbelt/local/lib/python2.7/site-packages/setuptools/tests/test_easy_install.py, line 157
      @mock.patch('setuptools.command.easy_install.__file__', None)
      def test_user_install_implied(self):
        fixture 'user_override' not found
        available fixtures: record_xml_property, recwarn, betamax_recorder, tmpdir, distutils_package, cache, monkeypatch, pytestconfig, setup_context, tmpdir_factory, betamax_session, capsys, capfd
        use 'py.test --fixtures [testpath]' for help on them.

/Users/ryan/src/requests-toolbelt/local/lib/python2.7/site-packages/setuptools/tests/test_easy_install.py:157
_________ ERROR at setup of TestUserInstallTest.test_multiproc_atexit __________
file /Users/ryan/src/requests-toolbelt/local/lib/python2.7/site-packages/setuptools/tests/test_easy_install.py, line 169
      def test_multiproc_atexit(self):
        fixture 'user_override' not found
        available fixtures: record_xml_property, recwarn, betamax_recorder, tmpdir, distutils_package, cache, monkeypatch, pytestconfig, setup_context, tmpdir_factory, betamax_session, capsys, capfd
        use 'py.test --fixtures [testpath]' for help on them.

/Users/ryan/src/requests-toolbelt/local/lib/python2.7/site-packages/setuptools/tests/test_easy_install.py:169
 ERROR at setup of TestUserInstallTest.test_user_install_not_implied_without_usersite_enabled 
file /Users/ryan/src/requests-toolbelt/local/lib/python2.7/site-packages/setuptools/tests/test_easy_install.py, line 180
      def test_user_install_not_implied_without_usersite_enabled(self):
        fixture 'user_override' not found
        available fixtures: record_xml_property, recwarn, betamax_recorder, tmpdir, distutils_package, cache, monkeypatch, pytestconfig, setup_context, tmpdir_factory, betamax_session, capsys, capfd
        use 'py.test --fixtures [testpath]' for help on them.

/Users/ryan/src/requests-toolbelt/local/lib/python2.7/site-packages/setuptools/tests/test_easy_install.py:180
____________ ERROR at setup of TestUserInstallTest.test_local_index ____________
file /Users/ryan/src/requests-toolbelt/local/lib/python2.7/site-packages/setuptools/tests/test_easy_install.py, line 190
      def test_local_index(self):
        fixture 'user_override' not found
        available fixtures: record_xml_property, recwarn, betamax_recorder, tmpdir, distutils_package, cache, monkeypatch, pytestconfig, setup_context, tmpdir_factory, betamax_session, capsys, capfd
        use 'py.test --fixtures [testpath]' for help on them.

/Users/ryan/src/requests-toolbelt/local/lib/python2.7/site-packages/setuptools/tests/test_easy_install.py:190
__________ ERROR at setup of TestUserInstallTest.test_setup_requires ___________
file /Users/ryan/src/requests-toolbelt/local/lib/python2.7/site-packages/setuptools/tests/test_easy_install.py, line 248
      def test_setup_requires(self):
        fixture 'user_override' not found
        available fixtures: record_xml_property, recwarn, betamax_recorder, tmpdir, distutils_package, cache, monkeypatch, pytestconfig, setup_context, tmpdir_factory, betamax_session, capsys, capfd
        use 'py.test --fixtures [testpath]' for help on them.

/Users/ryan/src/requests-toolbelt/local/lib/python2.7/site-packages/setuptools/tests/test_easy_install.py:248
________ ERROR at setup of TestEggInfo.test_egg_base_installed_egg_info ________
file /Users/ryan/src/requests-toolbelt/local/lib/python2.7/site-packages/setuptools/tests/test_egg_info.py, line 56
      def test_egg_base_installed_egg_info(self, tmpdir_cwd, env):
        fixture 'tmpdir_cwd' not found
        available fixtures: record_xml_property, recwarn, env, betamax_recorder, tmpdir, cache, monkeypatch, pytestconfig, tmpdir_factory, betamax_session, capsys, capfd
        use 'py.test --fixtures [testpath]' for help on them.

/Users/ryan/src/requests-toolbelt/local/lib/python2.7/site-packages/setuptools/tests/test_egg_info.py:56
___________ ERROR at setup of TestUploadDocsTest.test_create_zipfile ___________
file /Users/ryan/src/requests-toolbelt/local/lib/python2.7/site-packages/setuptools/tests/test_upload_docs.py, line 42
      def test_create_zipfile(self):
        fixture 'user_override' not found
        available fixtures: record_xml_property, recwarn, betamax_recorder, tmpdir, sample_project, cache, monkeypatch, pytestconfig, tmpdir_factory, betamax_session, capsys, capfd
        use 'py.test --fixtures [testpath]' for help on them.

/Users/ryan/src/requests-toolbelt/local/lib/python2.7/site-packages/setuptools/tests/test_upload_docs.py:42
=================================== FAILURES ===================================
__________________ TestDistInfo.test_conditional_dependencies __________________

self = <setuptools.tests.test_dist_info.TestDistInfo instance at 0x110c9fcf8>

    @pytest.mark.importorskip('ast')
    def test_conditional_dependencies(self):
        specs = 'splort==4', 'quux>=1.1'
        requires = list(map(pkg_resources.Requirement.parse, specs))
    
        for d in pkg_resources.find_distributions(self.tmpdir):
            assert d.requires() == requires[:1]
>           assert d.requires(extras=('baz',)) == requires
E           assert [Requirement....ra == "baz"')] == [Requirement.p...('quux>=1.1')]
E             At index 1 diff: Requirement.parse('quux>=1.1; extra == "baz"') != Requirement.parse('quux>=1.1')
E             Full diff:
E             + [Requirement.parse('splort==4'), Requirement.parse('quux>=1.1')]
E             - [Requirement.parse('splort==4'),
E             -  Requirement.parse('quux>=1.1; extra == "baz"')]

local/lib/python2.7/site-packages/setuptools/tests/test_dist_info.py:36: AssertionError
_____________________ TestPackageIndex.test_links_priority _____________________

self = <setuptools.tests.test_packageindex.TestPackageIndex instance at 0x11112d560>

    def test_links_priority(self):
        """
            Download links from the pypi simple index should be used before
            external download links.
            https://bitbucket.org/tarek/distribute/issue/163
    
            Usecase :
            - someone uploads a package on pypi, a md5 is generated
            - someone manually copies this link (with the md5 in the url) onto an
              external page accessible from the package page.
            - someone reuploads the package (with a different md5)
            - while easy_installing, an MD5 error occurs because the external link
              is used
            -> Setuptools should use the link from pypi, not the external one.
            """
        if sys.platform.startswith('java'):
            # Skip this test on jython because binding to :0 fails
            return
    
        # start an index server
        server = IndexServer()
        server.start()
        index_url = server.base_url() + 'test_links_priority/simple/'
    
        # scan a test index
        pi = setuptools.package_index.PackageIndex(index_url)
        requirement = pkg_resources.Requirement.parse('foobar')
        pi.find_packages(requirement)
        server.stop()
    
        # the distribution has been found
>       assert 'foobar' in pi
E       assert 'foobar' in <setuptools.package_index.PackageIndex object at 0x111b8f5d0>

local/lib/python2.7/site-packages/setuptools/tests/test_packageindex.py:125: AssertionError
----------------------------- Captured stderr call -----------------------------
127.0.0.1 - - [06/May/2016 22:37:04] code 404, message File not found
127.0.0.1 - - [06/May/2016 22:37:04] "GET /setuptools/tests/indexes/test_links_priority/simple/foobar/ HTTP/1.1" 404 -
Couldn't find index page for 'foobar' (maybe misspelled?)
127.0.0.1 - - [06/May/2016 22:37:04] code 404, message File not found
127.0.0.1 - - [06/May/2016 22:37:04] "GET /setuptools/tests/indexes/test_links_priority/simple/ HTTP/1.1" 404 -
____________________ test_options_passing_on_newer_requests ____________________

    @mock.patch.object(requests, '__build__', 0x020500)
    # @mock.patch.object(poolmanager, 'PoolManager')
    def test_options_passing_on_newer_requests():
        """Show that options are passed for a new enough version of requests."""
>       _AppEngineConnection
E       NameError: global name '_AppEngineConnection' is not defined

tests/test_appengine_adapter.py:16: NameError
3 failed, 201 passed, 11 skipped, 20 error in 5.15 seconds
ERROR: InvocationError: '/Users/ryan/src/requests-toolbelt/.tox/py27/bin/py.test'
_______________________________________ summary ________________________________________
ERROR:   py27: commands failed
```

--- a/requests_toolbelt/adapters/appengine.py
+++ b/requests_toolbelt/adapters/appengine.py
@@ -100,6 +100,11 @@ class _AppEngineConnection(object):
         # We once tried to verify our assumptions here, but sometimes the
         # passed-in URL differs on url fragments, or "http://a.com" vs "/".
 
+        # urllib3's App Engine adapter only uses Timeout.total, not read or
+        # connect.
+        if not timeout.total:
+            timeout.total = timeout._read or timeout._connect
+
         # Jump through the hoops necessary to call AppEngineManager's API.
         return self.appengine_manager.urlopen(
             method,

--- a/tests/test_appengine_adapter.py
+++ b/tests/test_appengine_adapter.py
@@ -1,0 +1,41 @@
+# -*- coding: utf-8 -*-
+"""Tests for the AppEngineAdapter."""
+import sys
+
+import mock
+import pytest
+import requests
+
+REQUESTS_SUPPORTS_GAE = requests.__build__ >= 0x021000
+
+if REQUESTS_SUPPORTS_GAE:
+    from requests.packages.urllib3.contrib import appengine as urllib3_appengine
+    from requests_toolbelt.adapters import appengine
+else:
+    appengine = urllib3_appengine = None
+
+
+@pytest.mark.skipif(sys.version_info >= (3,),
+                    reason="App Engine doesn't support Python 3 (yet)")
+@pytest.mark.skipif(not REQUESTS_SUPPORTS_GAE,
+                    reason="Requires Requests v2.10.0 or later")
+@mock.patch.object(urllib3_appengine, 'urlfetch')
+def test_get(mock_urlfetch):
+    """Tests a simple requests.get() call.
+
+    App Engine urlfetch docs:
+    https://cloud.google.com/appengine/docs/python/refdocs/google.appengine.api.urlfetch
+    """
+    response = mock.Mock(status_code=200, content='asdf', headers={})
+    mock_urlfetch.fetch = mock.Mock(return_value=response)
+
+    session = requests.Session()
+    session.mount('http://', appengine.AppEngineAdapter())
+    resp = session.get('http://url/', timeout=9, headers={'Foo': 'bar'})
+    assert resp.status_code == 200
+    assert resp.content == 'asdf'
+
+    args, kwargs = mock_urlfetch.fetch.call_args
+    assert args == ('http://url/',)
+    assert kwargs['deadline'] == 9
+    assert kwargs['headers']['Foo'] == 'bar'

